src/natural_history_manager.rs

```diff
@@ -0,0 +1,656 @@
+        // Let's first deal with the corner case -- the person is experiencing their first infection attempt.
+        // In this case, gi_cdf_value_unif will be 0.0. There are no points below 0.0 in a CDF, so interpolation
+        // will fail. Instead, we return 0.0. This is the obvious value because it means that the person
+        // has experienced none of their infectiousness at the start of their infection. It also ensures that if
+        // GI CDF is 0.0 for some time after the start of infection, inverse_gi(\epsilon) - inverse_gi(0) = c > 0
+        // even as \epsilon -> 0, which properly reproduces a CDF where an individual is not infectious immediately.
+        if gi_cdf_value_unif == 0.0 {
+            return 0.0;
+        }
+        // Grab the set GI trajectory for this person.
+        let natural_history_idx = self
+            .get_person_property(person_id, NaturalHistoryIdx)
+            .expect("No GI index set. Has this person been infected?")
+            .idx;
+        let natural_history_container = self
+            .get_data_container(NaturalHistory)
+            .expect("Natural history data container not initialized.");
+        let times = &natural_history_container.times[natural_history_idx];
+        let gi_trajectory = &natural_history_container.gi_trajectories[natural_history_idx];
+        // Because we want to interpolate the *inverse* CDF, the CDF values are "x" and the time values are "y".
+        interpolate(gi_trajectory, times, gi_cdf_value_unif)
+    }
+}
+
+fn assign_natural_history_idx(context: &mut Context, person_id: PersonId) {
+    if let Some(NaturalHistoryIdxValue { time, .. }) =
+        context.get_person_property(person_id, NaturalHistoryIdx)
+    {
+        // We've already assigned a natural history index for this person at this time.
+        if time == context.get_current_time() {
+            return;
+        }
+    }
+    let natural_history_len = context
+        .get_data_container(NaturalHistory)
+        .unwrap()
+        .gi_trajectories
+        .len();
+    let natural_history_idx =
+        context.sample_range(NaturalHistorySamplerRng, 0..natural_history_len);
+    context.set_person_property(
+        person_id,
+        NaturalHistoryIdx,
+        Some(NaturalHistoryIdxValue {
+            idx: natural_history_idx,
+            time: NotNan::new(context.get_current_time()).unwrap(),
+        }),
+    );
+}
+
+/// An interpolation routine that expects a paired set of values `xs` and `ys` that represent samples
+/// from a given function. The function is evaluated at a given x value `xp` using cubic spline interpolation
+/// when there are at least two samples above and below `xp`. Otherwise, it uses linear extrapolation at the tails.
+/// Assumes that function samples are sorted so that the `xs` are in ascending order.
+fn interpolate(xs: &[f64], ys: &[f64], xp: f64) -> f64 {
+    let window_right_edge_option = xs.iter().position(|&x| x > xp);
```

Comment:
I believe that `.iter().position()` does a linear scan. I would recommend doing binary search.

Moreover, if I understand the algorithm correctly, for a given agent each query will have a larger value of `xp` than the previous query in which case we could remember that position, thus saving more time going forward.